plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -130,6 +232,152 @@ class bp_connection_manager {
+            // peer key may have changed or been removed on-chain, do not consider that a fatal error, just remove it
+            i = msg.peers.erase(i);
+         } else {
+            ++i;
+         }
+      }
+
+      if (fatal_error)
+         return false;
+      return !msg.peers.empty();
+   }
+
+   // thread-safe
+   bool update_gossip_bps(const gossip_bp_peers_message& msg) {
+      // providing us with full set
+      fc::lock_guard g(gossip_bps.mtx);
+      auto& idx = gossip_bps.index.get<by_producer>();
+      bool diff = false;
+      for (const auto& peer : msg.peers) {
+         if (auto i = idx.find(boost::make_tuple(peer.producer_name, boost::cref(peer.server_address))); i != idx.end()) {
+            if (i->sig != peer.sig) { // signature has changed, producer_name and server_address has not changed
+               gossip_bps.index.modify(i, [&peer](auto& m) {
+                  m.sig = peer.sig; // update the signature, producer_name and server_address has not changed
+               });
+               diff = true;
+            }
+         } else {
+            if (idx.count(peer.producer_name) >= max_bp_peers_per_producer) {
+               // only allow max_bp_peers_per_producer, choose one to remove
+               gossip_bps.index.erase(idx.find(peer.producer_name));
+            }
+            gossip_bps.index.insert(peer);
+            diff = true;
+         }
+      }
+      return diff;
+   }
+
+   flat_set<std::string> find_gossip_bp_addresses(const peer_name_set_t& accounts, const char* desc) const {
+      flat_set<std::string> addresses;
+      fc::lock_guard g(gossip_bps.mtx);
+      const auto& prod_idx = gossip_bps.index.get<by_producer>();
+      for (const auto& account : accounts) {
+         if (auto i = config.bp_peer_addresses.find(account); i != config.bp_peer_addresses.end()) {
+            fc_dlog(self()->get_logger(), "${d} manual bp peer ${p}", ("d", desc)("p", i->second));
+            addresses.insert(i->second.address());
+         }
+         auto r = prod_idx.equal_range(account);
+         for (auto i = r.first; i != r.second; ++i) {
+            fc_dlog(self()->get_logger(), "${d} gossip bp peer ${p}", ("d", desc)("p", i->server_address));
+            addresses.insert(i->server_address);
+         }
+      }
+      return addresses;
+   }
+
+   // thread-safe
+   void connect_to_active_bp_peers() {
+      // do not hold mutexes when calling resolve_and_connect which acquires connections mutex since other threads
+      // can be holding connections mutex when trying to acquire these mutexes
+      flat_set<std::string> addresses;
+      {
+         fc::lock_guard gm(mtx);
+         active_bps = active_bp_accounts(active_schedule);
+         addresses = find_gossip_bp_addresses(active_bps, "connect");
+      }
+
+      for (const auto& add : addresses) {
+         self()->connections.resolve_and_connect(add, self()->get_first_p2p_address());
```

Comment:
This will likely fill the debug log with:

`Already connected to ${p}`

mabe pass a flag to `resolve_and_connect` to avoid logging if already connected.

Reply:
https://github.com/AntelopeIO/spring/pull/1265/commits/f778d714a3202f9647b37188c35ac8ea13af90db